asllib/ASLTypingReference.mld

```diff
@@ -16,107 +16,753 @@ welcome; those can be sent to Arm’s Architecture Formal Team Lead Jade
+
+    func main() => integer
+    begin
+      var dataT1: T1;
+      var pair: pairT = (1, dataT1);
+
+      let dataAsInt: integer = dataT1;
+      pair = (1, dataAsInt);
+      return 0;
+    end
+    ]}
+    [pair = (1, dataAsInt);] is legal since the right-hand-side has anonymous,
+    primitive type [(integer, integer)].
+ 
+    {4 TypingRule.TypeSatisfaction3.asl}
+    In the program:
+    {@asl[
+    type T1 of integer;          // the named type `T1` whose structure is integer
+    type T2 of integer;          // the named type `T2` whose structure is integer
+    type pairT of (integer, T1); // the named type `pairT` whose structure is (integer, integer)
+
+    func main() => integer
+    begin
+      var dataT1: T1;
+      var pair: pairT = (1,dataT1);
+
+      let dataT2: T2 = 10;
+      pair = (1, dataT2);
+      return 0;
+    end
+    ]}
+    [pair = (1, dataT2);] is illegal since the right-hand-side has anonymous,
+    non-primitive type [(integer, T2)] which does not subtype-satisfy named
+    type [pairT].
+
+{2 TypingRule.TypeClash}
+
+  {3 Definition}
+  [T] type-clashes with [S] if one of the following applies:
+  - [S] and [T] both have the structure of integers;
+  - [S] and [T] both have the structure of reals;
+  - [S] and [T] both have the structure of strings;
+  - [S] and [T] both have the structure of enumeration types with the same enumeration literals;
+  - [S] and [T] both have the structure of bitvectors;
+  - [S] and [T] both have the structure of arrays whose element types type-clash;
+  - [S] and [T] both have the structure of tuples of the same length whose
+    corresponding element types type-clash;
+  - [S] is a subtype of [T];
+  - [S] is a supertype of [T].
+
+  {3 Comments}
+  Note that if T subtype-satisfies S then T and S type-clash, but not the other
+  way around.
+
+  Note that type-clashing is an equivalence relation. Therefore if T
+  type-clashes with A and B then it is also the case that A and B type-clash.
+
+  This encompasses LRM Section 7.3.3. D_VPZZ, I_PQCT and I_WZKM.
+
+  {3 Examples}
+
+{2 TypingRule.LowestCommonAncestor}
+
+  {3 Definition}
+  The lowest common ancestor of types S and T is ty and one of the following applies:
+  - All of the following applies:
+    - S and T are the same type;
+    - ty is S.
+
+  - All of the following applies:
+    - S and T are both named types;
+    - ty is a common supertype of S and T;
+    - ty is a subtype of all other common supertypes of S and T.
+
+  - All of the following applies:
+    - S and T both have the structure of array types with the same index type
+      and the same element types;
+    - One of the following applies:
+      - All of the following applies:
+        - S is a named type;
+        - T is an anonymous type;
+        - ty is S.
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is a named type;
+        - ty is T.
```

Comment:
I think this misses the case where S and T are anonymous.

Maybe we want to factor out all the common anonymous cases?